File: ZKCache/src/main/java/com/ericsson/ema/tim/dml/predicate/Eq.java
package com.ericsson.ema.tim.dml.predicate;

import com.ericsson.ema.tim.dml.DataTypes;
import com.ericsson.ema.tim.dml.DmlBadSyntaxException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public class Eq extends AbstractPredicate implements Predicate {
    private final static Logger LOGGER = LoggerFactory.getLogger(Eq.class);

    private Eq(String field, String value) {
        super(field, value);
    }

    public static Eq eq(String field, String value) {
        return new Eq(field, value);
    }

    @Override
    public boolean eval(Object tuple) {
        if (this.valueToComp == null)
            return true;

        Object fieldVal = getFiledValFromTupleByName(tuple);
        Map<String, String> metadata = getSelector().getContext().getTableMetadata();
        String fieldType = metadata.get(field);
        switch (fieldType) {
            case DataTypes.String:
                return (this.valueToComp).equals(fieldVal);
            case DataTypes.Int:
                return (Integer.valueOf((String) this.valueToComp)).equals(fieldVal);
            default:
                LOGGER.error("unsupported data type: {},{}", field, fieldType);
                throw new DmlBadSyntaxException("unsupported data type: " + field + "," + fieldType);
        }
    }
}

File: zkcache_scala/src/main/scala/com/ericsson/ema/tim/dml/predicate/Eq.scala
package com.ericsson.ema.tim.dml.predicate


import com.ericsson.ema.tim.dml.DataTypes
import com.ericsson.ema.tim.exception.{DmlBadSyntaxException, DmlNoSuchFieldException}
import org.slf4j.LoggerFactory

/**
  * Created by eqinson on 2017/5/12.
  */
class Eq private(override val field: String, override val valueToComp: Object) extends AbstractPredicate(field, valueToComp) with Predicate {
	private val LOGGER = LoggerFactory.getLogger(classOf[Eq])

	override def eval(tuple: Object): Boolean = {
		if (this.valueToComp eq null)
			return false

		val fieldVal: Object = getFiledValFromTupleByName(tuple)
		val fieldType = selector.context.tableMetadata.get(field)
		fieldType match {
			case Some(DataTypes.String)  =>
				this.valueToComp == fieldVal
			case Some(DataTypes.Int)     =>
				Integer.valueOf(this.valueToComp.asInstanceOf[String]) == fieldVal.asInstanceOf[Integer]
			case Some(DataTypes.Boolean) =>
				java.lang.Boolean.valueOf(this.valueToComp.asInstanceOf[String]) == fieldVal.asInstanceOf[java.lang.Boolean]
			case Some(other)             =>
				LOGGER.error("bug: unsupported data type: {},{}", field, other: Any)
				throw DmlBadSyntaxException("unsupported data type: " + field + "," + other)
			case None                    => throw DmlNoSuchFieldException(field)
		}
	}
}

object Eq {
	def apply(field: String, value: String) = new Eq(field, value)
}
